index.js

```diff
@@ -0,0 +1,214 @@
+    typeof body === 'object' && body !== null,
+    400,
+    'Request body should be an object'
+  )
+  httpAssert(
+    typeof body.scores === 'object' && body.scores !== null,
+    400,
+    '.scores should be an object'
+  )
+  httpAssert(
+    Object.keys(body.scores).every(ethers.isAddress),
+    400,
+    'All .scores keys should be 0x addresses'
+  )
+  httpAssert(
+    Object.values(body.scores).every(score => {
+      try {
+        return BigInt(score) > 0n
+      } catch {
+        return false
+      }
+    }),
+    400,
+    'All .scores values should be positive numbers encoded as string'
+  )
+
+  validateSignature(body.signature, body.scores, signerAddresses)
+
+  const timestamp = new Date()
+  const tx = redis.multi()
+  for (const [address, score] of Object.entries(body.scores)) {
+    const scheduledRewards = (BigInt(score) * roundReward) / maxScore
+    tx.hincrby('rewards', address, scheduledRewards)
+    tx.rpush(
+      'log',
+      JSON.stringify({
+        timestamp,
+        address,
+        score,
+        scheduledRewards: String(scheduledRewards)
+      })
+    )
+  }
+  const updated = await tx.exec()
+
+  json(
+    res,
+    Object.fromEntries(
+      Object
+        .keys(body.scores)
+        .map((address, i) => [address, String(updated[i * 2][1])])
+    )
+  )
+}
+
+async function handlePaidScheduledRewards (req, res, redis, signerAddresses) {
+  const body = JSON.parse(await getRawBody(req, { limit: '1mb' }))
+
+  httpAssert(
+    typeof body === 'object' && body !== null,
+    400,
+    'Request body should be an object'
+  )
+  httpAssert(
+    typeof body.rewards === 'object' && body.rewards !== null,
+    400,
+    '.rewards should be an object'
+  )
+  httpAssert(
+    Object.keys(body.rewards).every(ethers.isAddress),
+    400,
+    'All .rewards keys should be 0x addresses'
+  )
+  httpAssert(
+    Object.values(body.rewards).every(score => {
+      try {
+        return BigInt(score) > 0n
+      } catch {
+        return false
+      }
+    }),
+    400,
+    'All .rewards values should be positive numbers encoded as string'
+  )
+
+  validateSignature(body.signature, body.rewards, signerAddresses)
+
+  const timestamp = new Date()
+  const tx = redis.multi()
+  for (const [address, amount] of Object.entries(body.rewards)) {
+    tx.hincrby('rewards', address, BigInt(amount) * -1n)
+    tx.rpush(
+      'log',
+      JSON.stringify({
+        timestamp,
+        address,
+        scheduledRewards: String(BigInt(amount) * -1n)
+      })
+    )
+  }
+  const updated = await tx.exec()
+
+  json(
+    res,
+    Object.fromEntries(
+      Object
+        .keys(body.rewards)
+        .map((address, i) => [address, String(updated[i * 2][1])])
+    )
+  )
+}
+
+async function handleGetAllScheduledRewards (res, redis) {
+  json(res, await redis.hgetall('rewards'))
+}
+
+async function handleGetSingleScheduledRewards (req, res, redis) {
+  json(res, await redis.hget('rewards', req.url.split('/').pop()))
```

Comment:
Some funky edge cases to consider:

These will return rewards (I think):

```
GET /scheduled-rewards/hello-world/0x000000000000000000000000000000000000dead
GET /scheduled-rewards?address=/0x000000000000000000000000000000000000dead
```

These will fail (I think):

```
GET /scheduled-rewards/0x000000000000000000000000000000000000dead/
GET /scheduled-rewards/0x000000000000000000000000000000000000dead?nocache=true
GET /scheduled-rewards/0x000000000000000000000000000000000000dead/?nocache=true
```

Reply:
Can we fix this once it becomes a problem? I'm trying to find a lean solution for now